src/examples/librpmem/fibonacci/fibonacci.c

```diff
@@ -0,0 +1,403 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * fibonacci.c -- fibonacci sequence generator for librpmem
+ */
+#include <assert.h>
+#include <errno.h>
+#include <time.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <libpmem.h>
+#include <librpmem.h>
+
+#define POOL_SIGNATURE	"FIBO"
+#define FIBO_SIG_LEN RPMEM_POOL_HDR_SIG_LEN
+#define SIZE_1MB	(1 * 1024 * 1024)
+
+struct fibo_t {
+	char signature[FIBO_SIG_LEN];
+	unsigned n;	/* index */
+	uint64_t fn;	/* F_{n} */
+	uint64_t fn1;	/* F_{n + 1} */
+	int checksum;
+};
+
+#define POOL_SIZE	(size_t)(32 * 1024 * 1024)
+#define DATA_OFF	4096 /* rpmem header size */
+#define NLANES		4
+#define DATA_SIZE	(sizeof(struct fibo_t))
+
+#define GET_FIBO(pool)	((struct fibo_t *)(pool + DATA_OFF))
+
+#define BROKEN_LOCAL	(1 << 0)
+#define BROKEN_REMOTE	(1 << 0)
+
+static void parse_args(int argc, char *argv[], const char **target,
+		const char **poolset, const char **path);
+static void *map_pmem(const char *path, size_t *mapped_len);
+
+static RPMEMpool *remote_create_or_open(const char *target, const char *poolset,
+		void *pool, int *created);
+static int remote_write(RPMEMpool *rpp);
+static int remote_read(RPMEMpool *rpp, void *buff);
+
+static void fibo_cleanup(RPMEMpool *rpp, void *pool, struct fibo_t *fibo_r,
+		int *initialized);
+static void fibo_generate(void *pool);
+static void fibo_print(void *pool);
+
+int
+main(int argc, char *argv[])
+{
+	const char *target, *poolset, *path;
+	parse_args(argc, argv, &target, &poolset, &path);
+
+	struct fibo_t fibo_r; /* copy from remote */
+	int created; /* remote: 1 == created, 0 == opened */
+	int initialized; /* sequence: 1 == initialized, 0 = con be continued */
+	int ret;
+
+	/* map local pool */
+	size_t mapped_len;
+	void *pool = map_pmem(path, &mapped_len);
+
+	/* open remote pool */
+	RPMEMpool *rpp = remote_create_or_open(target, poolset, pool, &created);
+	if (!rpp) {
+		ret = 1;
+		goto unmap;
+	}
+
+	if (!created) {
+		/* restore from remote */
+		ret = remote_read(rpp, &fibo_r);
+		if (ret)
+			/* invalidate remote copy */
+			memset(&fibo_r, 1, DATA_SIZE);
+	} else
+		/* zero remote copy */
+		memset(&fibo_r, 0, DATA_SIZE);
+
+	/* validate copies from local and remote */
+	fibo_cleanup(rpp, pool, &fibo_r, &initialized);
+	/* generate new number */
+	if (!initialized)
+		fibo_generate(pool);
+	fibo_print(pool);
+
+	/* store to remote */
+	ret = remote_write(rpp);
+	if (ret)
+		goto err;
+
+	printf("rerun application to generate next Fibonacci number.\n");
+
+err:
+	/* close the remote pool */
+	if (rpmem_close(rpp)) {
+		fprintf(stderr, "rpmem_close: %s\n", rpmem_errormsg());
```

Comment:
I closing failed can we even unmap the pool?

Reply:
Yes, we can. The memory is unregistered from the RDMA so it can be released. The potential issue can come from an SSH connection and an obstacle with sending a close command to rpmemd.

Reply:
As a caller of rpmem_close you don't know why it failed...

Reply:
I agree this API is not perfect. ;-)

Assuming we have no idea what happened inside we can just call ```exit(1)```. Do you think it is a better solution?

Reply:
I think it is a better solution.